### PR TITLE
Fix CI timeouts and windows posix errors

### PR DESCRIPTION
**Pull request by 0candy**
- Fix timeout errors
- Fix windows posix `test` command
- Fix mysql on windows by default table names are lower case.

Connect to https://github.com/strongloop/loopback-workspace/issues/322

### CODE REVIEW

**Comment by jannyHou** [comment]
I assume this pr is created to fix failings reported here :)
https://cis-jenkins.swg-devops.com/job/ds/job/loopback-workspace~master/62/console

My PR in generator-loopback has similar problem on all node 0.10 platforms:
One test case fails due to timeout, then others fail due to 
> Error: ENOTEMPTY, rmdir some_file_path

I thought it's a node 0.10 problem but seems the CI link I post above uses node 4.x.
Also I could not reproduce it on local with node 4.4.7, for example if I run two failing tests together but comment `done()` in the first one, the second one still pass, no ENOTEMPTY error.
I will try with 0.10.

**Comment by 0candy** [comment]
@jannyHou I seem to get different results each time. However, it passes on my local with node 0.10

**Comment by jannyHou** [comment]
Link issue https://github.com/strongloop/loopback-workspace/pull/312

**Comment by bajtos** [comment]
I am almost certainly sure that if workspace fails on Node v0.10 only, then it's caused by https://github.com/rxaviers/cldr-data-npm/pull/34

**Comment by bajtos** [comment]
FWIW, rxaviers/cldr-data-npm#34 has been landed & released, I think this patch should not be needed any more. For example, CI build of https://github.com/strongloop/loopback-workspace/pull/312 is passing on Node v0.10 AFAICT.

@jannyHou @0candy

**Comment by bajtos** [comment]
However, there seem to be other timing-related build failures on Node v0.12, see http://ci.strongloop.com/job/loopback-workspace/label=amazon-0.12/4757/consoleText

```

  1) end-to-end notes template "before all" hook:
     Error: timeout of 10000ms exceeded. Ensure the done() callback is being called in this test.
  

  2) end-to-end hello-world template "before all" hook:
     Error: EEXIST, mkdir '/mnt/jenkins/workspace/loopback-workspace/1de32925/test/sandbox'
      at Error (native)

  3) end-to-end start/stop/restart restarts the app:
     Error: timeout of 10000ms exceeded. Ensure the done() callback is being called in this test.
      at null.<anonymous> (node_modules/loopback/node_modules/loopback-context/node_modules/continuation-local-storage/node_modules/async-listener/glue.js:188:31)

  4) Facet Facet.create(def, cb) "before each" hook for "should use name as the id":
     Error: timeout of 10000ms exceeded. Ensure the done() callback is being called in this test.
      at null.<anonymous> (node_modules/loopback/node_modules/loopback-context/node_modules/continuation-local-storage/node_modules/async-listener/glue.js:188:31)
```

**Comment by 0candy** [comment]
@slnode test please

**Comment by 0candy** [comment]
I reran it and now it fails for node 6. I think there is still a timing issue not dependent on node version.

File: package.json [hunk 1]
```diff
@@ -6,7 +6,7 @@
     "export-tests": true
   },
   "scripts": {
-    "pretest": "test -z \"$JENKINS_HOME\" || (echo | node bin/setup-mysql.js)",
```
Comment (rmg) [inline package.json]: The `echo |` is still needed because `node bin/setup-mysql.js` is prompting for a password.

Comment (superkhau) [inline package.json]: Will that work on Windows?

Comment (rmg) [inline package.json]: We can try it and find out. The alternative is to modify `bin/setup-mysql.js` so that it doesn't use `read()`.

Comment (0candy) [inline package.json]: `echo |` does not work for windows.

**Comment by rmg** [comment]
Perhaps the `bin/setup-mysql.js` script should accept a `--auto` flag that makes it not prompt for a password and always exit successfully. If run _without_ the `--auto` flag, it could use the current behaviour of prompting for a password and exiting with failure if appropriate.

**Comment by bajtos** [comment]
Let's step back and ask ourselves what is the `pretest` hook trying to achieve?

 - When the developer is running `npm test` locally, the SQL setup should be skipped. We assume they have already set up their database.
 - When running `npm test` on a CI server, we may have a fresh SQL server instance and need to setup the database & access credentials. The assumption is that the CI server has an admin user with an empty password.

In that light, I think we should change `bin/setup-mysql.js` to detect CI, skip `read` and use an empty password.

```js
function askForPassword(next) {
  if (process.env.CI) {
    password = '';
    return next();
  }
  read({
    prompt: g.f('Enter password for MySQL root user:'),
    silent: true,
  }, function(err, pwd) {
    if (err) return next(err);
    password = pwd;
    next();
  });
}
```

**Comment by bajtos** [comment]
Moving here the discussion from https://github.com/strongloop/loopback-workspace/issues/322:

> If the script is relatively non-intrusive and developers don't mind it running every time they do npm test then it could be made unconditional.
> We should just remove the jenkins check altogether.

I disagree.

The script needs MySQL `admin` password to work correctly. Having to type the password every time I run `npm test` is rather intrusive to me personally.

**Comment by superkhau** [comment]
>When running npm test on a CI server, we may have a fresh SQL server instance and need to setup the database & access credentials. The assumption is that the CI server has an admin user with an empty password.

Can the CI instance be set up to use MySQL default user (ie. root/nopass)? External users do not have access to CI anyways.

@bajtos Can we just write the script to assume root/nopass since that is how default MySQL instances come OOB anyways? Then leave it up to the user running `npm test` to set up their env however they want (whether it be CI or community, etc).

**Comment by rmg** [comment]
> Can the CI instance be set up to use MySQL default user (ie. root/nopass)? External users do not have access to CI anyways.

@superkhau that _is_ how it is set up, and how this script is able to run. The problem is `bin/setup-mysql.js` is written in such a way that it essentially *requires* someone to "press return" instead of a non-interactive mechanism.

> When the developer is running `npm test` locally, the SQL setup should be skipped. We assume they have already set up their database.

I personally prefer modules to be a _little_ more turn-key and idempotent than that. Can we at least having `npm test` tell me (as a new contributor) what I need to do to get `npm test` to pass if I run it before running `bin/setup-mysql.js`?

> In that light, I think we should change bin/setup-mysql.js to detect CI, skip read and use an empty password.

FYI, the connectors use a different convention and instead check for and use `MYSQL_USER`/`TEST_MYSQL_USER`/etc env vars for their credentials. It doesn't make a difference to me, I just want to make sure we are aware of a pseudo-convention that is already in place.

Whichever approach is taken, I strongly agree that it needs to be non-interactive.

**Comment by superkhau** [comment]
>The problem is bin/setup-mysql.js is written in such a way that it essentially requires someone to "press return" instead of a non-interactive mechanism.
>Whichever approach is taken, I strongly agree that it needs to be non-interactive.

I agree, if we're going to require the script be run, there should be an option for non-interactive mode.

**Comment by bajtos** [comment]
> FYI, the connectors use a different convention and instead check for and use MYSQL_USER/TEST_MYSQL_USER/etc env vars for their credentials. It doesn't make a difference to me, I just want to make sure we are aware of a pseudo-convention that is already in place.

IIRC, I decided to not use those env vars because they are rather clumsy to configure on a local dev machine. It may also cause problems when there are multiple projects sharing the same database, let's say loopback-workspace and loopback-connector-mysql, especially when the connector tests require the database to be in a certain initial state IIRC.

I don't mind changing this decision, as long as we have an easy way how to setup local dev env.

However, I think my proposal in https://github.com/strongloop/loopback-workspace/pull/311#issuecomment-247989528 is the fastest way how to get CI green on Windows.

> I personally prefer modules to be a _little_ more turn-key and idempotent than that. Can we at least having `npm test` tell me (as a new contributor) what I need to do to get `npm test` to pass if I run it before running `bin/setup-mysql.js`?

The test suite is already printing an useful message when it cannot connect to MySQL, see https://github.com/strongloop/loopback-workspace/blob/3e80fcd3053199c5de549a5ded5e7773ad2fc5cc/test/end-to-end.js#L1059-L1085

```js
  connection.connect(function(err) {
    if (!err) return done(err);
    if (err.code === 'ECONNREFUSED') {
      err = new Error(
          'Cannot connect to local MySQL database, ' +
          'make sure you have `mysqld` running on your machine');
    } else {
      console.error();
      console.error('**************************************');
      console.error('Cannot connect to MySQL.');
      console.error('Setup the test environment by running');
      console.error('    node bin/setup-mysql');
      console.error('**************************************');
      console.error();
    }
    done(err);
  });
```

**Comment by superkhau** [comment]
I vote for @bajtos suggestion at https://github.com/strongloop/loopback-workspace/pull/311#issuecomment-247989528 to get Windows green now.

**Comment by rmg** [comment]
The reason for using env vars from CI instead of hard-coding `''` is so that no code has to be modified if CI needs to change credentials or any other aspect of DB connectivity.

Ideally every part of the connection string would be pulled from env vars, but from your proposal it wouldn't take much to extend it to use and env var with a fallback. eg:
```js
if (process.env.CI) {
  password = process.env.MYSQL_PASSWORD || '';
  return next();
}
```

**Comment by rmg** [comment]
@slnode test please

File: test/end-to-end.js [hunk 2]
```diff
@@ -29,8 +29,15 @@ var MYSQL_DATABASE = 'loopback_workspace_test';
 var MYSQL_USER = 'lbws';
 var MYSQL_PASSWORD = 'hbx42rec';
 
+before(function cleanTestDir(cb) {
+  fs.remove(PKG_CACHE, function(err) {
```
Comment (bajtos) [inline test/end-to-end.js]: Ouch, this completely cancels the point of using `strong-cached-install`. If we are going to delete `PKG_CACHE` before the tests run, then it's probably better to use `npm install` directly without `strong-cached-install`.

What problem is this trying to solve? Also how can be `PKG_CACHE` non-empty on CI machines, when each CI run starts in a clean cloned/checked-out repository?

I introduced `strong-cached-install` to speed up `npm test` when it's run repeatedly on a developer's machine. I.e. the first `npm test` is pretty slow, but subsequent `npm test` runs are much faster, as `npm install` is replaced with recursive copy.

Can we revert this change?

File: package.json [hunk 3]
```diff
@@ -31,6 +31,7 @@
     "morgan": "^1.2.0",
     "ncp": "^2.0.0",
     "normalize-git-url": "^3.0.2",
+    "path": "^0.12.7",
```
Comment (raymondfeng) [inline package.json]: Why an external `path` dep?

File: bin/setup-mysql.js [hunk 4]
```diff
@@ -17,23 +17,31 @@ USER = 'lbws';
 PASSWORD = 'hbx42rec';
 
 var connection;
+var password = '';
```
Comment (bajtos) [inline bin/setup-mysql.js]: nitpick: no need to initialise the variable here.

```js
var password, connection;
```

File: test/facet.js [hunk 5]
```diff
@@ -8,6 +8,9 @@ var Facet = require('../').models.Facet;
 
 describe('Facet', function() {
   describe('Facet.create(def, cb)', function() {
+    this.timeout(15000);
+
+    before(resetWorkspace);
     beforeEach(givenBasicWorkspace);
```
Comment (bajtos) [inline test/facet.js]: Isn't `givenBasicWorkspace` calling `resetWorkspace` under the hood? I find this change little bit suspicious, but if it fixes CI failures then I don't object.

File: test/model-definition.js [hunk 6]
```diff
@@ -13,6 +13,7 @@ var ModelRelation = app.models.ModelRelation;
 var TestDataBuilder = require('./helpers/test-data-builder');
 var ref = TestDataBuilder.ref;
 var ConfigFile = app.models.ConfigFile;
+var nodePath = require('path');
```
Comment (bajtos) [inline test/model-definition.js]: This is confusing, (almost all) Node.js code uses `var path = require('path')`. Can you please rename the other variables that are clashing with the `path` name?

File: test/model-definition.js [hunk 7]
```diff
@@ -125,8 +126,9 @@ describe('ModelDefinition', function() {
       expect(path).to.equal(configFilePath);
     });
     it('should return construct configFile path', function() {
+      var modelPath = 'models/my-model.json';
       var path = ModelDefinition.getPath('.', { name: 'MyModel' });
```
Comment (bajtos) [inline test/model-definition.js]: Use e.g. `modelPath` here so that `path` can be used for `require('path')`.

**Comment by 0candy** [comment]
Create new issue for the remaining windows failures for 404 errors.
https://github.com/strongloop/loopback-workspace/issues/328

**Pull request merged**